### PR TITLE
Centaur test JM-like metadata requests before and after carboniting [BA-6124]

### PR DESCRIPTION
**Pull request by grsterin**

### CODE REVIEW

File: centaur/src/main/scala/centaur/test/formulas/TestFormulas.scala [hunk 1]
```diff
@@ -22,6 +23,35 @@ import spray.json.JsString
...
+      "executionEvents",
+      "labels",
+      "parentWorkflowId",
+      "returnCode",
+      "status",
+      "submission",
+      "subWorkflowId",
+      "workflowName"
+    ),
+    "excludeKey" -> List("callCaching:hitFailures")
```
Comment (mcovarr) [inline centaur/src/main/scala/centaur/test/formulas/TestFormulas.scala]: I wouldn't expect any of the keys with `:` in them to work in Carbonited metadata the way they do with "classic", fixing that is in the vein of BA-6087.

File: centaur/src/main/scala/centaur/test/Test.scala [hunk 2]
```diff
@@ -539,6 +539,41 @@ object Operations extends StrictLogging {
     }
   }
 
+  private def checkDiff(diffs: Iterable[String],
+                        actualMetadata: WorkflowMetadata,
+                        errorMessage: String,
+                        submittedWorkflow: SubmittedWorkflow,
+                        workflowSpec: Workflow): Option[CentaurTestException] = {
```
[the remark this answers is not in the document]
Comment (aednichols) [reply centaur/src/main/scala/centaur/test/Test.scala]: Isn't an `IO` designed to indicate I/O or other side-effecting computation? Constructing and storing an exception does seem like more of an `Option` to me.

File: centaur/src/main/scala/centaur/test/metadata/WorkflowFlatMetadata.scala [hunk 3]
```diff
@@ -24,14 +24,18 @@ import scala.util.{Failure, Success, Try}
...
+  def diff(actual: WorkflowFlatMetadata, workflowID: UUID, cacheHitUUID: Option[UUID] = None, compareTyped: Boolean = false): Iterable[String] = {
     // If the test fails in initialization there wouldn't be workflow root metadata, and if that's the expectation
     // then that's ok.
     val workflowRoot = actual.value.get("workflowRoot").collectFirst { case JsString(r) => r } getOrElse "No Workflow Root"
     val missingErrors = value.keySet.diff(actual.value.keySet) map { k => s"Missing key: $k" }
-    val mismatchErrors = value.keySet.intersect(actual.value.keySet) flatMap { k => diffValues(k, value(k), actual.value(k),
-      workflowID, workflowRoot, cacheHitUUID)}
-
+    val mismatchErrors = value.keySet.intersect(actual.value.keySet) flatMap { k =>
+      if (compareTyped) {
```
Comment (aednichols) [inline centaur/src/main/scala/centaur/test/metadata/WorkflowFlatMetadata.scala]: [TOL] Personal preference, I find creating separate functions easier to follow than introducing a highly significant flag in an existing one
```
def diff(actual: WorkflowFlatMetadata, workflowID: UUID, cacheHitUUID: Option[UUID] = None) = {
  val (workflowRoot, missingErrors) = checkPreconditions(actual)
  // call to `diff`
}
def diffCheckTypes(actual: WorkflowFlatMetadata, workflowID: UUID, cacheHitUUID: Option[UUID] = None) = {
  val (workflowRoot, missingErrors) = checkPreconditions(actual)
  // call to `diffValuesTyped`
}

private def checkDiffPreconditions(actual: WorkflowFlatMetadata): (String, Set[String]) = {
  val workflowRoot: String = actual.value.get("workflowRoot").collectFirst { case JsString(r) => r } getOrElse "No Workflow Root"
  val missingErrors: Set[String] = value.keySet.diff(actual.value.keySet) map { k => s"Missing key: $k" }

  (workflowRoot, missingErrors)
}
```

Comment (grsterin) [reply centaur/src/main/scala/centaur/test/metadata/WorkflowFlatMetadata.scala]: @aednichols, @cjllanwarne, @mcovarr on a second thought, I'm not sure, should we really introduce a test case (Centaur or unit) for just one specific combination of input parameters? Maybe we should just cover the specific cases (like semicolons in metadata keys) by fine-grained unit tests?

Running comparisons between maybe-not-yet-carbonited metadata and carbonited metadata in each Centaur test seems like a perfect way to make our long-running tests to run even longer 🙂

Comment (mcovarr) [reply centaur/src/main/scala/centaur/test/metadata/WorkflowFlatMetadata.scala]: my 2 cents: I'm okay with closing this for the reasons you state.

**Pull request closed without merging**